Mathlib/Data/Fintype/BigOperators.lean

```diff
@@ -0,0 +1,306 @@
+#align fintype.prod_extend_by_one Fintype.prod_extend_by_one
+#align fintype.sum_extend_by_zero Fintype.sum_extend_by_zero
+
+end
+
+section
+
+variable {M : Type _} [Fintype α] [CommMonoid M]
+
+@[to_additive]
+theorem prod_eq_one (f : α → M) (h : ∀ a, f a = 1) : (∏ a, f a) = 1 :=
+  Finset.prod_eq_one fun a _ha => h a
+#align fintype.prod_eq_one Fintype.prod_eq_one
+#align fintype.sum_eq_zero Fintype.sum_eq_zero
+
+@[to_additive]
+theorem prod_congr (f g : α → M) (h : ∀ a, f a = g a) : (∏ a, f a) = ∏ a, g a :=
+  Finset.prod_congr rfl fun a _ha => h a
+#align fintype.prod_congr Fintype.prod_congr
+#align fintype.sum_congr Fintype.sum_congr
+
+@[to_additive]
+theorem prod_eq_single {f : α → M} (a : α) (h : ∀ (x) (_ : x ≠ a), f x = 1) : (∏ x, f x) = f a :=
+  Finset.prod_eq_single a (fun x _ hx => h x hx) fun ha => (ha (Finset.mem_univ a)).elim
+#align fintype.prod_eq_single Fintype.prod_eq_single
+#align fintype.sum_eq_single Fintype.sum_eq_single
+
+@[to_additive]
+theorem prod_eq_mul {f : α → M} (a b : α) (h₁ : a ≠ b) (h₂ : ∀ x, x ≠ a ∧ x ≠ b → f x = 1) :
+    (∏ x, f x) = f a * f b := by
+  apply Finset.prod_eq_mul a b h₁ fun x _ hx => h₂ x hx <;>
+    exact fun hc => (hc (Finset.mem_univ _)).elim
+#align fintype.prod_eq_mul Fintype.prod_eq_mul
+#align fintype.sum_eq_add Fintype.sum_eq_add
+
+/-- If a product of a `Finset` of a subsingleton type has a given
+value, so do the terms in that product. -/
+@[to_additive "If a sum of a `Finset` of a subsingleton type has a given
+  value, so do the terms in that sum."]
+theorem eq_of_subsingleton_of_prod_eq {ι : Type _} [Subsingleton ι] {s : Finset ι} {f : ι → M}
+    {b : M} (h : (∏ i in s, f i) = b) : ∀ i ∈ s, f i = b :=
+  Finset.eq_of_card_le_one_of_prod_eq (Finset.card_le_one_of_subsingleton s) h
+#align fintype.eq_of_subsingleton_of_prod_eq Fintype.eq_of_subsingleton_of_prod_eq
+#align fintype.eq_of_subsingleton_of_sum_eq Fintype.eq_of_subsingleton_of_sum_eq
+
+end
+
+end Fintype
+
+open Finset
+
+section
+
+variable {M : Type _} [Fintype α] [CommMonoid M]
+
+@[to_additive (attr := simp)]
+theorem Fintype.prod_option (f : Option α → M) : (∏ i, f i) = f none * ∏ i, f (some i) :=
+  Finset.prod_insertNone f univ
+#align fintype.prod_option Fintype.prod_option
+#align fintype.sum_option Fintype.sum_option
+
+end
+
+open Finset
+
+@[simp]
+nonrec theorem Fintype.card_sigma {α : Type _} (β : α → Type _) [Fintype α] [∀ a, Fintype (β a)] :
+    Fintype.card (Sigma β) = ∑ a, Fintype.card (β a) :=
+  card_sigma _ _
+#align fintype.card_sigma Fintype.card_sigma
+
+@[simp]
+theorem Finset.card_pi [DecidableEq α] {δ : α → Type _} (s : Finset α) (t : ∀ a, Finset (δ a)) :
+    (s.pi t).card = ∏ a in s, card (t a) :=
+  Multiset.card_pi _ _
+#align finset.card_pi Finset.card_pi
+
+@[simp]
+theorem Fintype.card_piFinset [DecidableEq α] [Fintype α] {δ : α → Type _} (t : ∀ a, Finset (δ a)) :
+    (Fintype.piFinset t).card = ∏ a, card (t a) := by simp [Fintype.piFinset, card_map]
```

Comment:
This is a mis-port; in Lean3 the `card` on the right is `finset.card`, here it's `Fintype.card`.